ui/app/components/transformation-edit.js

```diff
@@ -64,28 +46,51 @@ export default TransformBase.extend({
           errorStatus: e.httpStatus,
         };
       });
+    // if an error occurs while querying the role, exit function and return the error
+    if (roleStore.errorStatus) return roleStore;
+    // otherwise update the role with the transformation and save
+    let transformations = roleStore.transformations;
+    if (role.action === 'ADD') {
+      transformations = addToList(transformations, transformationId);
+    } else if (role.action === 'REMOVE') {
+      transformations = removeFromList(transformations, transformationId);
+    }
+    roleStore.setProperties({
+      backend,
+      transformations,
+    });
+    return roleStore.save().catch((e) => {
+      return {
+        errorStatus: e.httpStatus,
+        ...role,
+      };
+    });
   },
 
   handleUpdateRoles(updateRoles, transformationId) {
     if (!updateRoles) return;
-    const backend = this.model.backend;
-    const promises = updateRoles.map((r) => this.updateOrCreateRole(r, transformationId, backend));
-
-    Promise.all(promises).then((results) => {
-      const hasError = results.find((role) => !!role.errorStatus);
-
-      if (hasError) {
-        let message =
-          'The edits to this transformation were successful, but transformations for its roles was not edited due to a lack of permissions.';
-        if (results.find((e) => !!e.errorStatus && e.errorStatus !== 403)) {
-          // if the errors weren't all due to permissions show generic message
-          // eg. trying to update a role with empty array as transformations
-          message = `You've edited the allowed_roles for this transformation. However, the corresponding edits to some roles' transformations were not made`;
-        }
-        this.flashMessages.info(message, {
-          sticky: true,
-          priority: 300,
-        });
+    const { backend } = this.model;
+    updateRoles.forEach(async (record) => {
+      // For each role that needs to be updated, update the role with the transformation.
+      const updateOrCreateResponse = await this.updateOrCreateRole(record, transformationId, backend);
+      // If an error was returned, check error type and show a message.
+      const errorStatus = updateOrCreateResponse?.errorStatus;
+      if (errorStatus == 403) {
+        this.flashMessages.info(
+          `The edits to this transformation were successful, but transformations for the role ${record.id} were not edited due to a lack of permissions.`,
+          {
+            sticky: true,
+            priority: 300,
+          }
+        );
+      } else if (errorStatus) {
+        this.flashMessages.info(
+          `You've edited the allowed_roles for this transformation. However, there was a problem updating the role: ${record.id}.`,
+          {
+            sticky: true,
+            priority: 300,
+          }
+        );
```

Comment:
Since we're not returning here, I wonder if this makes double flash messages appear? I think we can keep it to `let message` so we're only calling this.flashMessages once

Reply:
The let cannot be defined inside the if statement and accessed in the else statement (hence the double flash message). But let me move the let outside the if all together and put flow back.